docs/architecture/adr-012-handlers-redesign.md

```diff
@@ -0,0 +1,598 @@
+# ADR 012: Handlers validation and execution separation
+
+## Status
+Proposed
+
+## Changelog
+* 9/9/2022: initial proposal
+
+## Context
+
+The main purpose of this ADR is to split each IBC handler (e.g. `UpdateClient`, `ConnOpenTry`, etc.) into a "validation" and an "execution" phase in order to accomodate a larger class of host architectures such as Namada. We call "validation" the part of a handler that does all the checks necessary for correctness. We call "execution" the part of the handler that applies the state modifications, assuming that validation passed. 
+
+Our current `deliver()` entrypoint can then be split into 2 entrypoints: `validate()` and `execute()`. More specifically, we replace the current `Ics26Context` and all its supertraits with 2 traits: `ValidationContext` and `ExecutionContext`. `ValidationContext` exposes and implements the `validate()` entrypoint, while the `ExecutionContext` exposes and implements the `execute()` entrypoint.
+
+This ADR will only concern itself with the external-facing API.
+
+The following are out of scope for this ADR:
++ Exposing an `async` API
++ Enabling light clients to define and require the host to implement a set of traits
+
+
+## Decision
+
+### "validation" and "execution" clarification
+We first clarify what is meant by "validation" and "execution" with an example. Consider the [`UpdateClient` handler](https://github.com/cosmos/ibc/blob/main/spec/core/ics-002-client-semantics/README.md#update). The validation and execution blocks are:
+
+```javascript
+function updateClient(
+  id: Identifier,
+  clientMessage: ClientMessage) {
+    // Validation START
+    clientState = provableStore.get(clientStatePath(id))
+    abortTransactionUnless(clientState !== null)
+
+    clientState.VerifyClientMessage(clientMessage)
+    // Validation END
+    
+    // Execution START
+    foundMisbehaviour := clientState.CheckForMisbehaviour(clientMessage)
```

Comment:
I think `CheckForMisbehaviour` should also be part of the `validation` here.

Reply:
Why do you say so? How I see it, it is not validation because it cannot make the execution fail. Rather, it is a `boolean` that you get to inform "how to do the execution": `UpdateStateOnMisbehaviour` or `UpdateState`.